README.md
# rtmpSnoop - The RTMP sniffer!

**rtmpSnoop** lets you to sniff RTMP streams from live TV, online channels and straming services and dump the RTMP properties in many formats.
You can analyse both live and dumped streams.

## Features

* Live sniffing from one ore more interfaces
* Read dumped streams from PCAP files
* Dump the RTMP properties in more formats (simple list, m3u entry or rtmpdump syntax)
* Easy to use and cross platform!

## Requirements

**rtmpSnoop** works both on Windows and Unix.  
To run it you need only python (at least 2.7 version) and the scapy module. 

**Linux Installation**  
* Debian/Ubuntu:  
`apt-get install python-scapy`

* RedHat/Centos:  
`yum install scapy.noarch`  
`yum install python-argparse.noarch`  

**Mac Installation**  

* Download pcapy from http://corelabs.coresecurity.com/
* Download dnet from http://libdnet.sourceforge.net/

Unzip and cd in to dnet file then

```
 CFLAGS='-arch i386 -arch x86_64' ./configure --prefix=/usr
 archargs='-arch i386 -arch x86_64' make
 sudo make install
 cd python
 sudo python setup.py install
```

**Windows Installation**  
Follow this guide to install scapy module on windows:
http://www.secdev.org/projects/scapy/doc/installation.html#windows

## Usage

The syntax is quite simple:

```
$python rtmpSnoop.py -h
usage: rtmpSnoop.py [-h] [-i DEVICE | -f PCAPFILE]
                    [--out-list | --out-m3u | --out-rtmpdump] [-p PORT]
                    [--one] [--quiet] [--debug]

rtmpSnoop lets you to grab the RTMP properties from live or dumped streams.

optional arguments:
  -h, --help      show this help message and exit

Input:
  -i DEVICE       Device to sniff on (Default: sniffs on all devices)
  -f PCAPFILE     PCAP file to read from

Output format:
  --out-list      Prints the RTMP data as list (Default)
  --out-m3u       Prints the RTMP data as m3u entry
  --out-rtmpdump  Prints the RTMP data in the rtmpdump format

Additional options:
  -p PORT         RTMP port (Default: sniffs on all ports)
  --one           Quit after the first stream found
  --quiet         Doesn't print anything except the RTMP output
  --debug         Enable DEBUG mode
```

## Examples

Sniffing on all interfaces, without filters:
```
sudo python rtmpSnoop.py
```

Sniffing on eth0, and looking for RTMP streams on port 1935 only:
```
sudo python rtmpSnoop.py -i eth0 -p 1935
```

Reading streams from PCAP file:
```
python rtmpSnoop.py -f dump/tv.pcap
```

## Output formats

Default list:
```
url: rtmp://192.168.1.1/live/channel?id=123
app: live
pageUrl: http://www.test.com/embedded/channel/1/500/380
swfUrl: http://www.test.eu/static/player.swf
tcUrl: rtmp://192.168.1.1/live
playPath: channel?id=123
flashVer: LNX 11,7,700,203
extra: S:OK 
```

m3u entry:
```
#EXTINF:0,1, Stream
rtmp://192.168.1.1/live/channel?id=12345 app=live pageUrl=http://www.test.eu/embedded/channel/1/500/380 
swfUrl=http://www.test.eu/static/player.swf tcUrl=rtmp://192.168.1.1/live playPath=channel?id=123 conn=S:OK live=1
```

rtmpdump syntax:
```
rtmpdump -r 'rtmp://192.168.1.1/live/channel?id=12345' -a 'live' -t 'rtmp://192.168.1.1/live' 
-y 'channel?id=12345' -W 'http://www.test.eu/scripts/player.swf' -p 'http://www.test.eu/embedded/channel/1/500/380' 
-f 'LNX 11,7,700,203' -C S:OK  --live -o stream.flv
```

## Donations

 If you want to support this project, please consider donating:
 * PayPal: andrea.fabrizi@gmail.com
 * BTC: 1JHCGAMpKqUwBjcT3Kno9Wd5z16K6WKPqG


CHANGELOG.md
# CHANGELOG

## Version 0.2 - Jan 16 2014:
* Fixed rtmpdump syntax bug
* Send log messages to stderr
* Fixed bug in quiet mode

## Version 0.1 - Jun 3 2013:
* Initial release 


lib/Utils.py
#!/usr/bin/env python
#
# This module is part of the rtmpSnoop project
#  https://github.com/andreafabrizi/rtmpSnoop
#
# Copyright (C) 2013 Andrea Fabrizi <andrea.fabrizi@gmail.com>
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public License
# as published by the Free Software Foundation; either version 3 of
# the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
# 02110-1301 USA
#

#Convert a number to a chr
def str2num(s):
    i = 0
    l = 0    
    while (i<len(s)):
        l = l << 8
        l = l + ord(s[i])
        i = i + 1

    return l


lib/Logger.py
#!/usr/bin/env python
#
# This module is part of the rtmpSnoop project
#  https://github.com/andreafabrizi/rtmpSnoop
#
# Copyright (C) 2013 Andrea Fabrizi <andrea.fabrizi@gmail.com>
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public License
# as published by the Free Software Foundation; either version 3 of
# the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
# 02110-1301 USA
#
import os, sys

class Logger():
    def __init__ (self, debug=False, quiet=False):
        self.DEBUG = debug
        self.QUIET = quiet

    def debug(self, message):
        if self.DEBUG and not self.QUIET:
            print >> sys.stderr, "# %s" % message

    def error(self, message):
        print >> sys.stderr, "*** %s" % message

    def info(self, message):
        if not self.QUIET:
            print >> sys.stderr, message


logger = Logger()


lib/Stream.py
#!/usr/bin/env python
#
# This module is part of the rtmpSnoop project
#  https://github.com/andreafabrizi/rtmpSnoop
#
# Copyright (C) 2013 Andrea Fabrizi <andrea.fabrizi@gmail.com>
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public License
# as published by the Free Software Foundation; either version 3 of
# the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
# 02110-1301 USA
#
from scapy.all import hexdump

class Stream():
    
    def __init__(self, stream = ""):
        self.stream = stream
        self.offset = 0
        self.size = len(stream)
        self.dontScanAgain = False

    #Appends new data to the stream
    def appendData(self, data):
        self.stream +=data
        self.size +=len(data)
    
    #Prints the stream 
    def dump(self):
        hexdump(self.stream[self.offset:])

    #Gets n bytes from the buffer and increments the offset
    def getBytes(self, n):
        if self.offset + n> self.size:
            raise StreamNoMoreBytes

        bytes = self.stream[self.offset:self.offset+n]
        self.offset = self.offset + n
        return bytes

    #Get a single byte from the stream
    def getByte(self):
        return ord(self.getBytes(1))

    #Reads n bytes from the buffer without increments the offset
    def readBytes(self, n):
        if self.offset >= self.size:
            return None

        bytes = self.stream[self.offset:self.offset+n]
        return bytes

    #Returns True if there are bytes to be read, False otherwise
    def haveBytes(self):
        if self.offset >= self.size:
            return False
        else:
            return True

class StreamNoMoreBytes(Exception):
    pass



lib/amfCommand.py
#!/usr/bin/env python
#
# This module is part of the rtmpSnoop project
#  https://github.com/andreafabrizi/rtmpSnoop
#
# Copyright (C) 2013 Andrea Fabrizi <andrea.fabrizi@gmail.com>
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public License
# as published by the Free Software Foundation; either version 3 of
# the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
# 02110-1301 USA
#
import os
from lib.Logger import logger

class amfCommand():

    def __init__(self):
        self.name = ""
        self.transaction_id = 0
        self.args = list()

            
class amfCommands():
    
    def __init__(self):
        self.commands = list()
        self.RTMP = dict()
        self.RTMP["extra"] = ""

    #Adds a new amfCommand object
    def add(self, amfCmd):
        self.commands.append(amfCmd)
    
    #Returns the amfCommand object by the name
    def get(self, name):
        for c in self.commands:
            if c.name == name:
                return c
                
    #Returns the number of the objects
    def count(self):
        return len(self.commands)
            
    #Prints the amf command object
    def parse(self):
        
        try:
        
            #Parsing the "connect" command arguments
            amfCmd = self.get("connect")
            
            for arg in amfCmd.args:
                if type(arg) == dict:
                    for prop in arg:
                        self.RTMP[prop] = arg[prop]
                else:
                
                    if type(arg) == str:
                        extra_type = "S:"
                    if type(arg) == bool:
                        extra_type = "B:"
                    if type(arg) == int:
                        extra_type = "N:"
                    
                    self.RTMP["extra"] += extra_type + str(arg) + " "

            #Parsing the "play" command arguments
            amfCmd = self.get("play")
            
            for arg in amfCmd.args:
                if arg:
                    self.RTMP["playPath"] = arg
                    break

            self.RTMP["url"] = os.path.join(self.RTMP["tcUrl"], self.RTMP["playPath"])
        
        except Exception as e:
            logger.error("Error during the RTMP properties parsing: %s" % e)

    def printBar(self):
        logger.info("*************************************")

    def printOut(self, mode):
        if mode == "m3u":
            self.printM3Uentry()
        elif mode == "rtmpdump":
            self.printRTMPDump()
        else:
            self.printDefault()

    """
    Prints the stream properties using the standard list format """
    def printDefault(self):
    
        self.parse()
        self.printBar()

        for prop in ["url", "app", "pageUrl", "swfUrl", "tcUrl", "playPath", "flashVer", "extra"]:
            if self.RTMP.has_key(prop) and len(self.RTMP[prop]) > 0:
                print "%s: %s" % (prop, self.RTMP[prop])

        self.printBar()

    """
    Prints out the RTMP properties using the m3u format """
    def printM3Uentry(self):

        self.parse()
        self.printBar()

        print "#EXTINF:0,1, Stream"

        line = "%s " % self.RTMP["url"]

        for prop in ["app", "pageUrl", "swfUrl", "tcUrl", "playPath"]:
            if self.RTMP.has_key(prop):
                line += "%s=%s " % (prop, self.RTMP[prop])

        if self.RTMP.has_key("extra"):
            line += "conn=%s" % self.RTMP["extra"]

        line += " live=1"

        print line
        self.printBar()

    """
    Prints out the RTMP properties using the rtmpdump format """
    def printRTMPDump(self):

        self.parse()
        self.printBar()

        line = "rtmpdump -r '%s' " % self.RTMP["url"]

        if self.RTMP["app"]:
            line += "-a '%s' " % self.RTMP["app"]
        else:
            line += "-a '' "

        if self.RTMP["tcUrl"]:
            line += "-t '%s' " % self.RTMP["tcUrl"]

        if self.RTMP["playPath"]:
            line += "-y '%s' " % self.RTMP["playPath"]

        if self.RTMP["swfUrl"]:
            line += "-W '%s' " % self.RTMP["swfUrl"]

        if self.RTMP["pageUrl"]:
            line += "-p '%s' " % self.RTMP["pageUrl"]

        if self.RTMP["flashVer"]:
            line += "-f '%s' " % self.RTMP["flashVer"]
        
        if self.RTMP["extra"]:
            line += "-C %s " % self.RTMP["extra"]

        line += "--live -o "
        if self.RTMP["playPath"]:
            filename = os.path.basename(self.RTMP["playPath"])
            if filename:
                line += filename
            else:
                line += "stream.flv"
        else:
            line += "stream.flv"
        
        print line
        self.printBar()


lib/rtmpParser.py
#!/usr/bin/env python
#
# This module is part of the rtmpSnoop project
#  https://github.com/andreafabrizi/rtmpSnoop
#
# Copyright (C) 2013 Andrea Fabrizi <andrea.fabrizi@gmail.com>
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public License
# as published by the Free Software Foundation; either version 3 of
# the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
# 02110-1301 USA
#
import struct
import Utils
from lib.Stream import Stream
from lib.amfCommand import amfCommand, amfCommands
from lib.Logger import logger
from scapy.all import *

class rtmpParser():

    AMF_COMMAND = 0x14
    AMF3_COMMAND = 0x11

    AMF_STRING = chr(0x02)
    AMF_NUMBER = chr(0x00)
    AMF_OBJECT = chr(0x03)
    AMF_BOOLEAN = chr(0x01)
    AMF_NULL = chr(0x05)
    AMF_ARRAY = chr(0x08)

    def __init__ (self):
        pass

    """ 
    Parses the stream packet by packet
    and returs a list containing all the amfCommand objects found """
    def rtmpParseStream (self, stream):

        amfCmds = amfCommands()
        
        #Looking for the handshake
        H1 = stream.getBytes(1)
        H1_rndData = stream.getBytes(0x600) #1536
        H2_rndData = stream.getBytes(0x600)

        if H1 != chr(0x03) or H1_rndData == None or H2_rndData == None:
            return amfCmds
    
        while (stream.haveBytes()):

            amf = self.rtmpParsePacket (stream)        

            if amf:
                amfCmds.add(amf)

        return amfCmds

    """
    Parses the RTMP object at the beginning of the stream.
    The stream pointer is incremented."""
    def rtmpParsePacket (self, stream):

        """
        Looking for the packet header
        byte 1
        BB BBBBBB
        The first 2 bit indicates the header type as following:
            b00 = 12 byte header (full header).
            b01 = 8 bytes - like type b00. not including message ID (4 last bytes).
            b10 = 4 bytes - Basic Header and timestamp (3 bytes) are included.
            b11 = 1 byte - only the Basic Header is included.

        The last 6 bit indicates the chunk stream ID
        """
        byte = stream.getByte()
        header_type = byte >> 6
        chunk_stream_id = byte << 2
        chunk_stream_id = chunk_stream_id >> 2

        #Header type b00
        if header_type == 0:
            timestamp = Utils.str2num(stream.getBytes(3))
            body_size = Utils.str2num(stream.getBytes(3))
            packet_type = stream.getByte()
            stream_id = Utils.str2num(stream.getBytes(4))
        
        #Header type b01
        elif header_type == 1:
            timestamp = Utils.str2num(stream.getBytes(3))
            body_size = Utils.str2num(stream.getBytes(3))
            packet_type = stream.getByte()

            """ 
        Header type b10 or b11
        Maybe this means that we reached the end of the stream
        with the setBufferLenght command, so simply I will set the entire stream "as readed"
        to end the parsing process """
        elif header_type == 2 or header_type == 3:
            stream.offset = stream.size
            return None

        else:
            logger.error("RTMP header type not supported: %d", header_type)
            return None

        #Now reading the RTMP payload from the stream
        magic_byte = 0xC0 + chunk_stream_id
        magic_bytes_count = body_size / 128
        rtmp_payload = stream.getBytes(body_size + magic_bytes_count)

        if rtmp_payload == None:
            return None

        #Unchunking the payload
        n = 0
        while (n<len(rtmp_payload)):
            if (n % 128 == 0) and (n != 0):
                if rtmp_payload[n] == chr(magic_byte):
                    rtmp_payload = rtmp_payload[:n] + rtmp_payload[n+1:]
                else:
                    logger.debug("Expected RTMP magic byte not found in the payload: %d" % n)
                    return None
            n = n + 1            
       
        """
        Now parsing the payload!
        I will create a new Stream object, containing only the payload, to pass to the
        parsing function """
        rtmp_payload_stream = Stream(rtmp_payload)

        #If it's an AMF/AMF3 command
        if packet_type == self.AMF_COMMAND or packet_type == self.AMF3_COMMAND:

            amf = amfCommand()

            #rtmp_payload_stream.dump()

            """ 
            In case of AMF3 command, there is an extra byte at the beginning of the body
            So, lets get it! """
            if packet_type == self.AMF3_COMMAND:
                rtmp_payload_stream.getByte()

            """
            The structure of the RTMP Command is:
                (String) <Command Name>
                (Number) <Transaction Id>
                (Mixed)  <Argument> ex. Null, String, Object: {key1:value1, key2:value2 ... }
            """

            #Reading AMF Command
            amf.name = self.rtmpParseObject(rtmp_payload_stream)

            #We are interested only in "connect" and "play" objects for our purpose
            if amf.name not in ["connect","play"]:
                logger.debug("Found an unuseful command, skypping!: %s" % amf.name)
                return None

            #Reading AMF Transaction ID
            amf.transaction_id = self.rtmpParseObject(rtmp_payload_stream)

            #Reading all the AMF arguments
            while (rtmp_payload_stream.haveBytes()):
                amf.args.append(self.rtmpParseObject(rtmp_payload_stream))

            return amf
        
        #Otherwise, I can discard it...
        else:
            logger.debug("Found an unuseful packet type, skypping!: %d" % packet_type)
            
        return None
        

    #Parse a single RTMP object
    def rtmpParseObject(self, p):

        #Object type
        b = p.getBytes(1)

        #STRING
        if (b == self.AMF_STRING):
            strlen =  Utils.str2num(p.getBytes(2))
            string = p.getBytes(strlen)
            logger.debug("Found a string [%s]..." % string)
            return string

        #NUMBER
        #Numbers are stored as 8 byte (big endian) float double
        elif (b == self.AMF_NUMBER):
            number = struct.unpack('>d',p.getBytes(8))
            logger.debug("Found a number [%d]..." % number)
            return int(number[0])

        #BOOLEAN
        elif (b == self.AMF_BOOLEAN):
            boolean = False if (p.getBytes(1) == chr(0)) else True
            logger.debug("Found a boolean (%s)..." % boolean)
            return boolean

        #OBJECT
        elif (b == self.AMF_OBJECT):
            logger.debug("Found an object...")
            obj = dict()

            #Reading all the object properties, until End Of Object marker is reached
            while (p.readBytes(3) != "\x00\x00\x09"):
                
                #Property name
                strlen =  Utils.str2num(p.getBytes(2))
                key = p.getBytes(strlen)
                logger.debug("Property name [%s]..." % key)

                #Property value
                val = self.rtmpParseObject(p)

                obj[key] = val

            #Eating the End Of Object marker
            p.getBytes(3)
            
            return obj

        #NULL
        elif (b == self.AMF_NULL):
            logger.debug("Found a NULL byte...")
            return None

        #ARRAY
        #I don't care about it...
        elif (b == self.AMF_ARRAY):
            arraylen =  Utils.str2num(p.getBytes(4))
            logger.debug("Found an array...")
            while (p.readBytes(3) != "\x00\x00\x09"):
                pass
            p.getBytes(3)
            return 0
       
        #Unknown object
        else:
            logger.error("Found an unknown RTMP object: 0x%x" % ord(b))
            return None



rtmpSnoop.py
#!/usr/bin/env python
#
# This module is part of the rtmpSnoop project
#  https://github.com/andreafabrizi/rtmpSnoop
#
# Copyright (C) 2013 Andrea Fabrizi <andrea.fabrizi@gmail.com>
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public License
# as published by the Free Software Foundation; either version 3 of
# the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
# 02110-1301 USA
#
import os
import logging
logging.getLogger("scapy.runtime").setLevel(logging.ERROR)
from scapy.all import *
from lib.Stream import *
from lib.rtmpParser import rtmpParser
from lib.amfCommand import amfCommand, amfCommands
from lib.Logger import logger
import argparse

VERSION="0.2.1"

"""
Packet Handler Callback from scapy
I don't care about the packet fragmentation, because is supposed that I'm sniffing on
a local interface, so shouldn't be a packet fragmentation problem.

"""
def PacketHandler(pkt):

    global streams
    global rtmp_port
    global out_mode
    global quit_first

    if pkt.haslayer(TCP) and pkt.haslayer(Raw):

        #Skipping if the rtmp_port is defined and is different from the packet dest port
        if rtmp_port != 0 and pkt[TCP].dport != rtmp_port:
            return

        sport = pkt[TCP].sport
        
        #hexdump(pkt.load)

        """
        The easiest way to follow the TCP streams is to use the source port as
        distinction element. So i will consider each packet with the same source
        port as part of the same TCP stream """
        if sport not in streams:
            stream = Stream(pkt.load)
            streams[sport] = stream
        else:
            streams[sport].appendData(pkt.load)

        if streams[sport].dontScanAgain:
            return
        
        #This is the mininium size that an RTMP stream must have to contains interesting data...
        if streams[sport].size > 0x600*2:
            logger.debug("Dissecting stream: %s" % sport)

            rtmp = rtmpParser()

            try:
                amfCmds = rtmp.rtmpParseStream(streams[sport])

                #If I have 2 AMF commands (play and connect), I can print the results
                if amfCmds.count() == 2:
                    logger.info("\n* RTMP Stream found!")
                    amfCmds.printOut(out_mode)
                    streams[sport].dontScanAgain = True
                    if quit_first:
                        sys.exit(0)
                else:
                    streams[sport].offset = 0

            except StreamNoMoreBytes:
                logger.debug("No more bytes to read from the stream!")

            except Exception as e:
                logger.error("Error parsing the RTMP stream: %s" % e)


"""
Setting up the argparse and usage """
def setupArgParser():

    parser = argparse.ArgumentParser(description="rtmpSnoop lets you to grab the RTMP properties from live or dumped streams.")

    group_input = parser.add_argument_group("Input")
    group = group_input.add_mutually_exclusive_group()
    group.add_argument("-i", action="store", dest="device", help="Device to sniff on (Default: sniffs on all devices)")
    group.add_argument("-f", action="store", dest="pcapfile", help="PCAP file to read from")

    group_output = parser.add_argument_group("Output format")
    group = group_output.add_mutually_exclusive_group()
    group.add_argument("--out-list", action='store_const', const="list", dest="out_mode", help="Prints the RTMP data as list (Default)")
    group.add_argument("--out-m3u", action='store_const', const="m3u", dest="out_mode", help="Prints the RTMP data as m3u entry")
    group.add_argument("--out-rtmpdump", action='store_const', const="rtmpdump", dest="out_mode", help="Prints the RTMP data in the rtmpdump format")

    group_input = parser.add_argument_group("Additional options")
    group_input.add_argument("-p", action="store", dest="port", default=0, type=int, help="RTMP port (Default: sniffs on all ports)")
    group_input.add_argument("--one", action="store_true", dest="quit_first", help="Quit after the first stream found")
    group_input.add_argument("--quiet", action="store_true", help="Doesn't print anything except the RTMP output")
    group_input.add_argument("--debug", action="store_true", help="Enable DEBUG mode")

    args = parser.parse_args()

    return args

#MAIN
if __name__ == "__main__":
    
    args = setupArgParser()
    
    if args.debug:
        logger.DEBUG = True

    if args.quiet:
        logger.QUIET = True

    rtmp_port = args.port
    out_mode = args.out_mode
    quit_first = args.quit_first

    logger.info("rtmpSnoop v%s - The RTMP Sniffer!" % VERSION)
    logger.info("Andrea Fabrizi - andrea.fabrizi@gmail.com\n")

    streams = dict()

    #Not sniffing, reading from dump file
    if args.pcapfile:
        logger.info("Reading packets from dump file '%s'..." % args.pcapfile)
        sniff(offline=args.pcapfile, filter="tcp", prn = PacketHandler)

    #Sniffing on the specified device
    elif args.device:
        logger.info("Starting sniffing on %s..." % args.device)
        try:
            sniff(iface=args.device, prn = PacketHandler)
        except socket.error as e:
            logger.error("Error opening %s for sniffing: %s" % (args.device, e))
            logger.info("Are you root and the device is correct?")
            sys.exit(1)

    #Default action, sniffing on all the devices
    else:
        logger.info("Starting sniffing on all devices...")
        try:
            sniff(prn = PacketHandler)
        except socket.error as e:
            logger.error("Error opening device for sniffing: %s" % e)
            logger.info("Are you root?")
            sys.exit(1)
    


